### ISSUE TITLE
Received maybe type parameters can't be used as maybe type arguments

### ISSUE THREAD

**Issue by dchambers** [facebook/flow#1508]
This code snippet causes an error:

~~~js
type Func = (obj?: Object) => Object;
const g: Func = (obj: ?Object): Object => obj || {};
const f = (obj: ?Object): Object => {
  return g(obj); // this line causes an error
};
~~~

whereas this code snippet (I've added ` || undefined`) works:

~~~js
type Func = (obj?: Object) => Object;
const g: Func = (obj: ?Object): Object => obj || {};
const f = (obj: ?Object): Object => {
  return g(obj || undefined);
};
~~~

**Comment by leonardfactory** [comment]
Declaring in the type `Func` first parameter as `obj?: Object` is different from `obj: ?Object`, even if in javascript it has a similar behaviour.

### TLDR
To do what I think you mean with your example, you should declare `Func` as: 

```js
type Func = (obj: ?Object) => Object;
```

You can see it [here working without errors](http://tryflow.org/#c2bf5b82464471ed1abcba14609be9eb).

### Further explanation

- `obj?: Object`: The parameter is optional, so you can call a `Func` with or without the parameter set, however if obj ___is passed___ it should be a **not-null** Object.

- `obj: ?Object`: The parameter _should_ be passed, however flow will not complain if you call only `g();` cause it's the same as `g(undefined);` which is allowed by the function declaration. 
  However in this case, `g(null);` is even allowed, so your function will not complain about a potentially `null` obj.

**Comment by dchambers** [comment]
Hi @leonardfactory, thanks so much for taking the time to explain where I've been going wrong &mdash; I'd come to the spurious conclusion that one form was used for functions while the other type was used for type definitions! The bonus of knowing that there's a semantic difference between these forms will no doubt come in handy down the line :smile:.

**Comment by dchambers** [comment]
In case it helps anyone else, it's worth pointing out that although you can both forms (`obj?: Object` and `obj: ?Object`) in _type-definitions_, you can only use the `obj: ?Object` form in _functions_ themselves.

Instead, when you need `obj? Object` within a function, you instead use the ES6 default parameter syntax (e.g. `obj: Object = {}`) to denote that the parameter is optional, and this then also saves you adding extra code within your function to deal with the fact that parameter may not have been defined.

**Comment by samwgoldman** [comment]
@dchambers I'm not sure that's accurate, unless I misunderstand. This function is perfectly fine:

```js
function f(x?: string): string {
  if (typeof x === "undefined") {
    return "default string";
  } else {
    return x;
  }
}
```

Now, you're correct that providing a default value has the effect of making the argument optional in the function type and non-optional in the method body, which is the natural behavior in my opinion. But it's totally fine to make an argument optional and deal with the possible `undefined` within the function.

**Comment by dchambers** [comment]
Hi @samwgoldman, maybe you're using a more recent version of Flow than I am, but it doesn't work for me, and doesn't work if [I modify the code in the demo link provided earlier](http://tryflow.org/#9203551ebe9b906cb62313f0bb1a00c7) either.

**Comment by dchambers** [comment]
BTW, the fact I couldn't do that is actually how I came to erroneous conclusion that one form was for _functions_ and the other for _type-definitions_ in the first place.

**Comment by samwgoldman** [comment]
Looks like a Babel bug, actually, and only affects optional arguments in arrow functions. Flow parses and types it successfully, but compilation fails. Will check with Babel folks to see if they're aware of this issue.

**Comment by dchambers** [comment]
Thanks @samwgoldman :+1:.

**Comment by samwgoldman** [comment]
Issue is being tracked here: https://phabricator.babeljs.io/T7096

**Comment by dchambers** [comment]
Thanks again for taking taking the time to look into this :smile:.
